burstphoto/align.swift

```diff
@@ -214,6 +724,196 @@ func texture_like(_ input_texture: MTLTexture) -> MTLTexture {
+    let max_threads_per_thread_group = state.maxTotalThreadsPerThreadgroup
+    let threads_per_thread_group = MTLSize(width: max_threads_per_thread_group, height: 1, depth: 1)
+    command_encoder.setTexture(in_texture, index: 0)
+    command_encoder.setTexture(avg_y, index: 1)
+    command_encoder.dispatchThreads(threads_per_grid, threadsPerThreadgroup: threads_per_thread_group)
+    
+    // average the generated 1d texture along the x-axis
+    let state2 = (pixelformat == "rgba" ? average_x_rgba_state : average_x_state)
+    command_encoder.setComputePipelineState(state2)
+    let avg_buffer = device.makeBuffer(length: (pixelformat=="rgba" ? 4 : 1)*MemoryLayout<Float32>.size, options: .storageModeShared)!
+    command_encoder.setTexture(avg_y, index: 0)
+    command_encoder.setBuffer(avg_buffer, offset: 0, index: 0)
+    command_encoder.setBytes([Int32(in_texture.width)], length: MemoryLayout<Int32>.stride, index: 1)
+    command_encoder.dispatchThreads(threads_per_grid, threadsPerThreadgroup: threads_per_thread_group)
+    command_encoder.endEncoding()
+    command_buffer.commit()
+    
+    // return the average of all pixels in the input array
+    return avg_buffer
+}
+
+
+
+// ===========================================================================================================
+// Functions specific to image alignment
+// ===========================================================================================================
+
+func align_texture(_ ref_pyramid: [MTLTexture], _ comp_texture: MTLTexture, _ downscale_factor_array: Array<Int>, _ tile_size_array: Array<Int>, _ search_dist_array: Array<Int>) -> MTLTexture {
+        
+    // initialize tile alignments
+    let alignment_descriptor = MTLTextureDescriptor.texture2DDescriptor(pixelFormat: .rg16Sint, width: 1, height: 1, mipmapped: false)
+    alignment_descriptor.usage = [.shaderRead, .shaderWrite]
+    var prev_alignment = device.makeTexture(descriptor: alignment_descriptor)!
+    var current_alignment = device.makeTexture(descriptor: alignment_descriptor)!
+    var tile_info = TileInfo(tile_size: 0, tile_size_merge: 0, search_dist: 0, n_tiles_x: 0, n_tiles_y: 0, n_pos_1d: 0, n_pos_2d: 0)
+    
+    // build comparison pyramid
+    let comp_pyramid = build_pyramid(comp_texture, downscale_factor_array)
+    
+    // align tiles
+    for i in (0 ... downscale_factor_array.count-1).reversed() {
+        
+        // load layer params
+        let tile_size = tile_size_array[i]
+        let search_dist = search_dist_array[i]
+        let ref_layer = ref_pyramid[i]
+        let comp_layer = comp_pyramid[i]
+        
+        // calculate the number of tiles
+        let n_tiles_x = ref_layer.width / (tile_size / 2) - 1
+        let n_tiles_y = ref_layer.height / (tile_size / 2) - 1
+        let n_pos_1d = 2*search_dist + 1
+        let n_pos_2d = n_pos_1d * n_pos_1d
+        
+        // store tile info in a single dict
```

Comment:
This is not a dict

Reply:
What would be the correct wording here?

Reply:
It's technically a struct, but the comment may be more useless if it says something like "storing the values together in a struct to it easier and more readable when passing between functions"